File: backend_ts/src/config/subjects.ts
export interface SubjectDefinition {
  code: string;
  name: string;
  isPractical: boolean;
}

export type CourseType = "BCA" | "BBA" | "BTECH" | "BSC" | "BCOM";

/**
 * Semester-wise subject definitions per course.
 * Structure: COURSE_SUBJECTS[course][semester] => SubjectDefinition[]
 */
export const COURSE_SUBJECTS: Record<string, Record<number, SubjectDefinition[]>> = {
  BCA: {
    1: [
      { code: "BCA0101", name: "Mathematics-I", isPractical: false },
      { code: "BCA0102", name: "Applied English", isPractical: false },
      { code: "BCA0103", name: "Computer Fundamentals", isPractical: false },
      { code: "BCA0104", name: "C Programming", isPractical: false },
      { code: "BCA0105", name: "Office Automation Tools", isPractical: false },
      { code: "BCA0104(P)", name: "C Programming Lab", isPractical: true },
      { code: "BCA0105(P)", name: "Office Automation Tools Lab", isPractical: true },
    ],
    2: [
      { code: "BCA0201", name: "Mathematics-II", isPractical: false },
      { code: "BCA0202", name: "Communicative English", isPractical: false },
      { code: "BCA0203", name: "Digital Electronics", isPractical: false },
      { code: "BCA0204", name: "Data Structures", isPractical: false },
      { code: "BCA0205", name: "Data Base Management System", isPractical: false },
      { code: "BCA0204(P)", name: "Data Structures Lab", isPractical: true },
      { code: "BCA0205(P)", name: "DBMS Lab", isPractical: true },
    ],
    3: [
      { code: "BCA0301", name: "Mathematics-III", isPractical: false },
      { code: "BCA0302", name: "Business Practices and Management", isPractical: false },
      { code: "BCA0303", name: "Object-Oriented Programming with C++", isPractical: false },
      { code: "BCA0304", name: "Desktop Publishing and Designing", isPractical: false },
      { code: "BCA0305", name: "Statistical Methods", isPractical: false },
      { code: "BCA0303(P)", name: "OOP with C++ Lab", isPractical: true },
      { code: "BCA0304(P)", name: "Desktop Publishing Lab", isPractical: true },
    ],
    4: [
      { code: "BCA0401", name: "Personnel Management", isPractical: false },
      { code: "BCA0402", name: "Accounting and Financial Management", isPractical: false },
      { code: "BCA0403", name: "System Analysis and Design", isPractical: false },
      { code: "BCA0404", name: "Internet Technology & Web Page Design", isPractical: false },
      { code: "BCA0405", name: "Programming in Visual Basic", isPractical: false },
      { code: "BCA0404(P)", name: "Web Page Design Lab", isPractical: true },
      { code: "BCA0405(P)", name: "Visual Basic Lab", isPractical: true },
    ],
    5: [
      { code: "BCA0501", name: "Operating System", isPractical: false },
      { code: "BCA0502", name: "Software Engineering", isPractical: false },
      { code: "BCA0503", name: "Object-Oriented Programming with Java", isPractical: false },
      { code: "BCA0504", name: "Computer Graphics", isPractical: false },
      { code: "BCA0503(P)", name: "Java Programming Lab", isPractical: true },
      { code: "BCA0504(P)", name: "Computer Graphics Lab", isPractical: true },
    ],
    6: [
      { code: "BCA0601", name: "Computer Networks", isPractical: false },
      { code: "BCA0602", name: "Numerical Methods", isPractical: false },
      { code: "BCA0603", name: "Multimedia Applications", isPractical: false },
      { code: "BCA0604", name: "Major Project", isPractical: false },
      { code: "BCA0605", name: "Seminar", isPractical: false },
      { code: "BCA0603(P)", name: "Multimedia Applications Lab", isPractical: true },
    ],
  },
  BBA: {
    1: [
      { code: "BBA101", name: "Principles of Management", isPractical: false },
      { code: "BBA102", name: "Business Communication", isPractical: false },
      { code: "BBA103", name: "Microeconomics", isPractical: false },
      { code: "BBA104", name: "Financial Accounting", isPractical: false },
      { code: "BBA105", name: "Business Mathematics", isPractical: false },
    ],
  },
  BTECH: {
    1: [
      { code: "BT101", name: "Engineering Mathematics", isPractical: false },
      { code: "BT102", name: "Engineering Physics", isPractical: false },
      { code: "BT103", name: "Engineering Chemistry", isPractical: false },
      { code: "BT104", name: "Programming Fundamentals", isPractical: false },
      { code: "BT102(P)", name: "Physics Lab", isPractical: true },
      { code: "BT103(P)", name: "Chemistry Lab", isPractical: true },
      { code: "BT104(P)", name: "Programming Lab", isPractical: true },
    ],
  },
};

/**
 * Get subjects for a given course and semester.
 * Falls back to an empty array if no config is found.
 */
export function getSubjectsForSemester(course: string, semester: number): SubjectDefinition[] {
  const courseConfig = COURSE_SUBJECTS[course.toUpperCase()];
  if (!courseConfig) return [];
  return courseConfig[semester] || [];
}

/**
 * Get all subjects across all semesters for a given course (flat list).
 */
export function getAllSubjectsForCourse(course: string): SubjectDefinition[] {
  const courseConfig = COURSE_SUBJECTS[course.toUpperCase()];
  if (!courseConfig) return [];
  const allSubjects: SubjectDefinition[] = [];
  for (const semester of Object.keys(courseConfig)) {
    const subs = courseConfig[Number(semester)];
    if (subs) {
      allSubjects.push(...subs);
    }
  }
  return allSubjects;
}

File: backend/Config/Subjects.cs
using System;
using System.Collections.Generic;

namespace EduGuard.Config
{
    public class SubjectDefinition
    {
        public string Code { get; set; } = string.Empty;
        public string Name { get; set; } = string.Empty;
        public bool IsPractical { get; set; }
    }

    public static class Subjects
    {
        public static readonly Dictionary<string, Dictionary<int, List<SubjectDefinition>>> CourseSubjects = new(StringComparer.OrdinalIgnoreCase)
        {
            ["BCA"] = new()
            {
                [1] = new()
                {
                    new() { Code = "BCA0101", Name = "Mathematics-I", IsPractical = false },
                    new() { Code = "BCA0102", Name = "Applied English", IsPractical = false },
                    new() { Code = "BCA0103", Name = "Computer Fundamentals", IsPractical = false },
                    new() { Code = "BCA0104", Name = "C Programming", IsPractical = false },
                    new() { Code = "BCA0105", Name = "Office Automation Tools", IsPractical = false },
                    new() { Code = "BCA0104(P)", Name = "C Programming Lab", IsPractical = true },
                    new() { Code = "BCA0105(P)", Name = "Office Automation Tools Lab", IsPractical = true }
                },
                [2] = new()
                {
                    new() { Code = "BCA0201", Name = "Mathematics-II", IsPractical = false },
                    new() { Code = "BCA0202", Name = "Communicative English", IsPractical = false },
                    new() { Code = "BCA0203", Name = "Digital Electronics", IsPractical = false },
                    new() { Code = "BCA0204", Name = "Data Structures", IsPractical = false },
                    new() { Code = "BCA0205", Name = "Data Base Management System", IsPractical = false },
                    new() { Code = "BCA0204(P)", Name = "Data Structures Lab", IsPractical = true },
                    new() { Code = "BCA0205(P)", Name = "DBMS Lab", IsPractical = true }
                },
                [3] = new()
                {
                    new() { Code = "BCA0301", Name = "Mathematics-III", IsPractical = false },
                    new() { Code = "BCA0302", Name = "Business Practices and Management", IsPractical = false },
                    new() { Code = "BCA0303", Name = "Object-Oriented Programming with C++", IsPractical = false },
                    new() { Code = "BCA0304", Name = "Desktop Publishing and Designing", IsPractical = false },
                    new() { Code = "BCA0305", Name = "Statistical Methods", IsPractical = false },
                    new() { Code = "BCA0303(P)", Name = "OOP with C++ Lab", IsPractical = true },
                    new() { Code = "BCA0304(P)", Name = "Desktop Publishing Lab", IsPractical = true }
                },
                [4] = new()
                {
                    new() { Code = "BCA0401", Name = "Personnel Management", IsPractical = false },
                    new() { Code = "BCA0402", Name = "Accounting and Financial Management", IsPractical = false },
                    new() { Code = "BCA0403", Name = "System Analysis and Design", IsPractical = false },
                    new() { Code = "BCA0404", Name = "Internet Technology & Web Page Design", IsPractical = false },
                    new() { Code = "BCA0405", Name = "Programming in Visual Basic", IsPractical = false },
                    new() { Code = "BCA0404(P)", Name = "Web Page Design Lab", IsPractical = true },
                    new() { Code = "BCA0405(P)", Name = "Visual Basic Lab", IsPractical = true }
                },
                [5] = new()
                {
                    new() { Code = "BCA0501", Name = "Operating System", IsPractical = false },
                    new() { Code = "BCA0502", Name = "Software Engineering", IsPractical = false },
                    new() { Code = "BCA0503", Name = "Object-Oriented Programming with Java", IsPractical = false },
                    new() { Code = "BCA0504", Name = "Computer Graphics", IsPractical = false },
                    new() { Code = "BCA0503(P)", Name = "Java Programming Lab", IsPractical = true },
                    new() { Code = "BCA0504(P)", Name = "Computer Graphics Lab", IsPractical = true }
                },
                [6] = new()
                {
                    new() { Code = "BCA0601", Name = "Computer Networks", IsPractical = false },
                    new() { Code = "BCA0602", Name = "Numerical Methods", IsPractical = false },
                    new() { Code = "BCA0603", Name = "Multimedia Applications", IsPractical = false },
                    new() { Code = "BCA0604", Name = "Major Project", IsPractical = false },
                    new() { Code = "BCA0605", Name = "Seminar", IsPractical = false },
                    new() { Code = "BCA0603(P)", Name = "Multimedia Applications Lab", IsPractical = true }
                }
            },
            ["BBA"] = new()
            {
                [1] = new()
                {
                    new() { Code = "BBA101", Name = "Principles of Management", IsPractical = false },
                    new() { Code = "BBA102", Name = "Business Communication", IsPractical = false },
                    new() { Code = "BBA103", Name = "Microeconomics", IsPractical = false },
                    new() { Code = "BBA104", Name = "Financial Accounting", IsPractical = false },
                    new() { Code = "BBA105", Name = "Business Mathematics", IsPractical = false }
                }
            },
            ["BTECH"] = new()
            {
                [1] = new()
                {
                    new() { Code = "BT101", Name = "Engineering Mathematics", IsPractical = false },
                    new() { Code = "BT102", Name = "Engineering Physics", IsPractical = false },
                    new() { Code = "BT103", Name = "Engineering Chemistry", IsPractical = false },
                    new() { Code = "BT104", Name = "Programming Fundamentals", IsPractical = false },
                    new() { Code = "BT102(P)", Name = "Physics Lab", IsPractical = true },
                    new() { Code = "BT103(P)", Name = "Chemistry Lab", IsPractical = true },
                    new() { Code = "BT104(P)", Name = "Programming Lab", IsPractical = true }
                }
            }
        };

        public static List<SubjectDefinition> GetSubjectsForSemester(string course, int semester)
        {
            if (string.IsNullOrEmpty(course)) return new();
            if (CourseSubjects.TryGetValue(course, out var semesters))
            {
                if (semesters.TryGetValue(semester, out var subs))
                {
                    return subs;
                }
            }
            return new();
        }

        public static List<SubjectDefinition> GetAllSubjectsForCourse(string course)
        {
            if (string.IsNullOrEmpty(course)) return new();
            var allSubjects = new List<SubjectDefinition>();
            if (CourseSubjects.TryGetValue(course, out var semesters))
            {
                foreach (var subs in semesters.Values)
                {
                    allSubjects.AddRange(subs);
                }
            }
            return allSubjects;
        }
    }
}
